release/releaser.go

```diff
@@ -45,539 +36,318 @@ func NewReleaser(
-			// TODO: Populate this... It's just included so the user can't trigger a null-pointer.
-			Result: flux.ReleaseResult{},
-		}, err)
-		if err2 != nil {
-			if err == nil {
-				err = errors.Wrap(err2, "sending notifications")
+			id, err := flux.ParseServiceID(string(s))
+			if err != nil {
+				return nil, err
 			}
+			ids = append(ids, id)
 		}
-	}()
-
-	for i, action := range actions {
-		updateJob(action.Description)
-		inst.Log("description", action.Description)
-		if action.Do == nil {
-			continue
+		return rc.SelectExactServices(ids, lockedSet, excludedSet, results, logStatus)
+	case flux.ServiceSpecAll:
+		return rc.SelectServices(nil, lockedSet, excludedSet, results, logStatus)
+	default:
+		id, err := flux.ParseServiceID(string(spec.ServiceSpec))
+		if err != nil {
+			return nil, err
 		}
+		return rc.SelectExactServices([]flux.ServiceID{id}, lockedSet, excludedSet, results, logStatus)
+	}
+}
 
-		if kind == flux.ReleaseKindExecute {
-			begin := time.Now()
-			result, err := action.Do(rc)
-			r.metrics.ActionDuration.With(
-				fluxmetrics.LabelAction, action.Name,
-				fluxmetrics.LabelSuccess, fmt.Sprint(err == nil),
-			).Observe(time.Since(begin).Seconds())
-			if err != nil {
-				updateJob(err.Error())
-				inst.Log("err", err)
-				actions[i].Result = "Failed: " + err.Error()
-				return err
-			}
-			if result != "" {
-				updateJob(result)
-			}
-			actions[i].Result = result
+// Find all the image updates that should be performed, and do
+// replacements. (For a dry-run, we don't strictly need to do the
+// replacements, since we won't be committing any changes back;
+// however we do want to see if we *can* do the replacements, because
+// if not, it indicates there's likely some problem with the running
+// system vs the definitions given in the repo.)
+func calculateImageUpdates(inst *instance.Instance, updateable []*ServiceUpdate, spec *flux.ReleaseSpec, results flux.ReleaseResult, logStatus statusFn) ([]*ServiceUpdate, error) {
+	// Compile an `ImageMap` of all relevant images
+	var images instance.ImageMap
+	var err error
+
+	switch spec.ImageSpec {
+	case flux.ImageSpecNone:
+		images = instance.ImageMap{}
+	case flux.ImageSpecLatest:
+		images, err = CollectAvailableImages(inst, updateable)
+	default:
+		var image flux.ImageID
+		image, err = spec.ImageSpec.AsID()
+		if err == nil {
+			images, err = inst.ExactImages([]flux.ImageID{image})
 		}
 	}
 
-	return nil
-}
+	if err != nil {
+		return nil, err
+	}
 
-func CalculateUpdates(services []platform.Service, images instance.ImageMap, printf func(string, ...interface{})) map[flux.ServiceID][]ContainerUpdate {
-	updateMap := map[flux.ServiceID][]ContainerUpdate{}
-	for _, service := range services {
-		containers, err := service.ContainersOrError()
+	// Look through all the services' containers to see which have an
+	// image that could be updated.
+	var updates []*ServiceUpdate
+	for _, update := range updateable {
```

Comment:
:+1:
Probably ought to stick to consistent names throughout, since many of these procedures have the same pattern of taking candidate updates in, and filtering them.